plugin/plugins/filesystem/nodes/ipfs.go

```diff
@@ -0,0 +1,204 @@
+
+	metaClone := id.meta
+	metaClone.Filter(req)
+
+	return id.Qid, req, metaClone, nil
+}
+
+func (id *IPFS) Walk(names []string) ([]p9.QID, p9.File, error) {
+	id.Logger.Debugf("ID Walk names %v", names)
+	id.Logger.Debugf("ID Walk myself: %s:%v", id.Path, id.Qid)
+
+	if doClone(names) {
+		id.Logger.Debugf("ID Walk cloned")
+		return []p9.QID{id.Qid}, id, nil
+	}
+
+	qids := make([]p9.QID, 0, len(names))
+
+	//TODO: [spec check] make sure we're not messing things up by doing this instead of mutating
+	// ^ Does internal library expect fid to mutate on success or does newfid clobber some external state anyway
+	walkedNode := &IPFS{} // operate on a copy
+	*walkedNode = *id
+
+	var err error
+	for _, name := range names {
+		//TODO: timerctx; don't want to hang forever
+		if walkedNode.Path, err = id.core.ResolvePath(id.Ctx, corepath.Join(walkedNode.Path, name)); err != nil {
+			return qids, nil, err
+		}
+
+		ipldNode, err := id.core.Dag().Get(id.Ctx, walkedNode.Path.Cid())
+		if err != nil {
+			return qids, nil, err
+		}
+
+		//TODO: this is too opague; we want core path => qid, dirent isn't necessary
+		dirEnt := &p9.Dirent{}
+		if err = ipldStat(dirEnt, ipldNode); err != nil {
+			return qids, nil, err
+		}
+
+		walkedNode.Qid = dirEnt.QID
+		//
+		qids = append(qids, walkedNode.Qid)
+	}
+
+	id.Logger.Debugf("ID Walk ret %v, %v", qids, walkedNode)
+	return qids, walkedNode, err
+}
+
+func (id *IPFS) Open(mode p9.OpenFlags) (p9.QID, uint32, error) {
+	id.Logger.Debugf("ID Open")
+	if id.meta.Mode.IsDir() {
+		c, err := id.core.Unixfs().Ls(id.Ctx, id.Path)
+		if err != nil {
+			return id.Qid, 0, err
+		}
+		id.directory = &directoryStream{
+			entryChan: c,
+		}
+	}
+
+	return id.Qid, 0, nil
+}
+
+func (id *IPFS) Readdir(offset uint64, count uint32) ([]p9.Dirent, error) {
+	id.Logger.Debugf("ID Readdir")
+
+	if id.directory == nil {
+		return nil, fmt.Errorf("directory %q is not open for reading", id.Path.String())
+	}
+
+	if count == 0 {
+		return nil, nil
+	}
+
+	if offset < id.directory.cursor {
+		return nil, fmt.Errorf("read offset %d is behind current entry %d, seeking backwards in directory streams is not supported", offset, id.directory.cursor)
+	}
+
+	ents := make([]p9.Dirent, 0)
+
+out:
+	for {
+		select {
+		case entry, open := <-id.directory.entryChan:
+			if !open {
+				break out
+			}
+			if entry.Err != nil {
+				return nil, entry.Err
+			}
+
+			if offset <= id.directory.cursor {
+				nineEnt := coreEntTo9Ent(entry)
+				nineEnt.Offset = id.directory.cursor
+				ents = append(ents, nineEnt)
+				count--
+			}
+
+			id.directory.cursor++
+
+			if count == 0 {
+				break out
+			}
+		case <-id.Ctx.Done():
+			return ents, id.Ctx.Err()
+		}
+	}
+
+	if offset > uint64(len(ents)) {
+		return nil, nil //cursor is at end of stream, nothing to return
+	}
+
+	id.Logger.Debugf("ID Readdir returning [%d]%v\n", len(ents), ents)
+	return ents, nil
+}
+
+func (id *IPFS) ReadAt(p []byte, offset uint64) (int, error) {
```

Comment:
The go docs for `ReaderAt` (https://golang.org/pkg/io/#ReaderAt) state that this should either fill `p` or return an error. However, I'm not sure if the library we're using requires that (or even wants that).

If we don't end up conforming to the ReadAt interface, we should document that.

Reply:
@Stebalien 
The commit for this seems to somehow gotten lost. Things have changed and I beleive we're conforming to the Go standard here.
When we encounter problems `p` is not modified, otherwise we defer to `go-ipfs-files.File` which implements the standard `io.Reader`

Can you look at this again? I might be missing something.
(also sorry for the huge delay here, Github swallowed this comment)